libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/GenState.hs

```diff
@@ -0,0 +1,478 @@
+  tlscript <- genNestedTimelock (2 :: Natural)
+  let corescript :: Core.Script era
+      corescript = case proof of
+        Babbage _ -> TimelockScript tlscript
+        Alonzo _ -> TimelockScript tlscript
+        Mary _ -> tlscript
+        Allegra _ -> tlscript
+        Shelley _ -> error "Shelley does not have TimeLock scripts"
+  let scriptHash = hashScript @era corescript
+  modify $ \ts@GenState {gsScripts} -> ts {gsScripts = Map.insert scriptHash corescript gsScripts}
+  pure scriptHash
+
+-- Adds to gsScripts
+genMultiSigScript :: forall era. Reflect era => Proof era -> GenRS era (ScriptHash (Crypto era))
+genMultiSigScript proof = do
+  let genNestedMultiSig k
+        | k > 0 =
+            elementsT $
+              nonRecTimelocks ++ [requireAllOf k, requireAnyOf k, requireMOf k]
+        | otherwise = elementsT nonRecTimelocks
+      nonRecTimelocks = [requireSignature]
+      requireSignature = Shelley.RequireSignature <$> genKeyHash
+      requireAllOf k = do
+        n <- lift nonNeg1Digit
+        Shelley.RequireAllOf <$> replicateM n (genNestedMultiSig (k - 1))
+      requireAnyOf k = do
+        n <- lift pos1Digit
+        Shelley.RequireAnyOf <$> replicateM n (genNestedMultiSig (k - 1))
+      requireMOf k = do
+        n <- lift nonNeg1Digit
+        m <- lift $ choose (0, n)
+        Shelley.RequireMOf m <$> replicateM n (genNestedMultiSig (k - 1))
+  msscript <- genNestedMultiSig (2 :: Natural)
+  let corescript :: Core.Script era
+      corescript = case proof of
+        Shelley _ -> msscript
+        _ -> error (show proof ++ " does not have MultiSig scripts")
+  let scriptHash = hashScript @era corescript
+  modify $ \ts@GenState {gsScripts} -> ts {gsScripts = Map.insert scriptHash corescript gsScripts}
+  pure scriptHash
+
+primaryLanguages :: Proof era -> [Language]
+primaryLanguages (Babbage _) = [PlutusV2]
+primaryLanguages (Alonzo _) = [PlutusV1]
+primaryLanguages _ = [PlutusV1]
+
+-- Adds to gsPlutusScripts
+genPlutusScript :: forall era. Reflect era => Proof era -> Tag -> GenRS era (ScriptHash (Crypto era))
+genPlutusScript proof tag = do
+  isValid <- lift $ frequency [(5, pure False), (95, pure True)]
+  -- Plutus scripts alwaysSucceeds needs at least numArgs, while
+  -- alwaysFails needs exactly numArgs to have the desired affect.
+  -- For reasons unknown, this number differs from Alonzo to Babbage
+  -- Perhaps because Babbage is using PlutusV2 scripts?
+  let numArgs = case (proof, tag) of
+        (Babbage _, Spend) -> 2
+        (Babbage _, _) -> 1
+        (_, Spend) -> 3
+        (_, _) -> 2
+  -- While using varying number of arguments for alwaysSucceeds we get
+  -- varying script hashes, which helps with the fuzziness
+  language <- lift $ elements (primaryLanguages proof)
+  script <-
+    if isValid
+      then -- alwaysSucceeds @era language . (+ numArgs) . getNonNegative <$> lift arbitrary
+        alwaysSucceeds @era language . (+ numArgs) <$> (lift $ elements [0, 1, 2, 3 :: Natural])
```

Comment:
Dead code and a way to avoid `$`
```suggestion
      then
        alwaysSucceeds @era language . (+ numArgs) <$> lift (elements [0, 1, 2, 3 :: Natural])
```